designdocs/provider_permissions.md

```diff
@@ -0,0 +1,205 @@
+
+### Goals
+
+**Opt In and Duality**: Providers should register users with MRC and delegate on behalf of them, while also specifically allowing a collection of schema(s) for which delegator provide them full publication rights. This ensures default state of providers is ***Restrict***. Delegators can also choose to restrict providers on per-schema basis by blocking them from publishing data on their behalf. This ensures default state of delegators is ***Block*** for all non provider preferred ```schema_ids```. Duality should be implemented at schema level grants.
+
+**ToS Baked In**: As a part of this design doc, it is recommended to discuss about baking in ***ToS*** for providers and delegators as a part of permission grants by including a hash of ToS unless there is a re-delegation. Such that MRC can also act as proof of specific agreement established between a provider and a delegator.
+
+**Time Bound Grants**: Any grants given or revoked by a delegator (allowing provider to publish or block them for certain duration) or any grants are modified by a provider or delegator are valid for the duration of ***ToS***. This can be a control mechanism in MRC which can be a fixed number for version 1 of this implementation and be extensible via a governance mechanism. This also brings the question about, if not time bounded, does permissions and grants are set till they are explicitly revoked. While un-delegation, definitely is an option for user to remove a provider completely from ever publishing on their behalf.
+
+### Non-Goals
+
+- Does not cover the case where a delegator or provider can restrict reading of data on their behalf.
+- MRC enables a valid provider or delegator to be able to read as a default.
+- Only covers basic version 1 of permission and grant implementation details.
+- Does not cover details of economics, governance mechanism.
+- Does not cover details on dynamic expiry time for permissions/grants.
+
+## Proposal
+
+The proposed solution is to provide delegate level permissions and schema level grants to delegators and providers alike. This will enable a provider to publish data on behalf of a delegator, while also allowing a delegator to restrict a provider from publishing data on their behalf for specific schema.
+
+Note: The terminology and implementation are subject to change at issue resolution.
+
+### Permissions
+
+Permission is a generic option for any user. For version 1 of this implementation, the following options are available:
+
+- ***PUBLISHER***: Where a user grants full publication rights to a provider for any schema available to provider via MRC. This can be modified to be called a dsnp publisher where all dsnp related schemas are granted to provider. In other sense this could be super admin permission that can be granted via a governance mechanism.
```

Comment:
I suggest we remove the Publisher option for version 1.

Why?
- Simplicity. The Restricted option can do everything
- User safety: Explicit user permission while annoying for upgrading to a different version, it protects against misuse. A grant could be given to someone who then creates a new ("bad" for the user) protocol and immediately start using it.

Reply:
Makes sense for version 1 we can have restricted and can extend things if needed